test/Constraints/sr2705.swift

```diff
@@ -0,0 +1,43 @@
+// RUN: %target-swift-emit-silgen %s -verify | %FileCheck %s
+
+func f<T>(_: () -> T) {}
+func f<T>(_: @autoclosure () -> T) {}
+
+// CHECK: function_ref @$s6sr27051fyyxyXElF : $@convention(thin) <τ_0_0> (@noescape @callee_guaranteed @substituted <τ_0_0> () -> @out τ_0_0 for <τ_0_0>) -> ()
+f { } // OK
+
+func f1<T>(_: () -> T, _: () -> T) {}
+func f1<T>(_: @autoclosure () -> T, _: @autoclosure () -> T) {}
+
+// CHECK: function_ref @$s6sr27052f1yyxyXE_xyXEtlF : $@convention(thin) <τ_0_0> (@noescape @callee_guaranteed @substituted <τ_0_0> () -> @out τ_0_0 for <τ_0_0>, @noescape @callee_guaranteed @substituted <τ_0_0> () -> @out τ_0_0 for <τ_0_0>) -> ()
+f1({}, {}) // OK
+
+func f2<T>(_: () -> T, _: () -> T) { }
+func f2<T>(_: () -> T, _: @autoclosure () -> T) { }
+
+// CHECK: function_ref @$s6sr27052f2yyxyXE_xyXEtlF : $@convention(thin) <τ_0_0> (@noescape @callee_guaranteed @substituted <τ_0_0> () -> @out τ_0_0 for <τ_0_0>, @noescape @callee_guaranteed @substituted <τ_0_0> () -> @out τ_0_0 for <τ_0_0>) -> () 
+f2({}, {}) // OK
+
+func f3(_: () -> Int) {}
+func f3(_: @autoclosure () -> Int) {}
+
+// CHECK: function_ref @$s6sr27052f3yySiyXEF : $@convention(thin) (@noescape @callee_guaranteed () -> Int) -> ()
```

Comment:
Nit: let's remove everything after `: ` in these checks, it's going to be much easier to fix later :)

Reply:
Ah ok so all type information? Just the function mangled name is enough to ensure the correct overload?

Reply:
Just fixed :)

Reply:
Great!

Reply:
@LucianoPAlmeida Yes, mangled name is enough since presence/absence of autoclosure is mangled in. There is a `swift-demangle` utility shipped with toolchain/compiler, you can use to print out actual type e.g.:

```
$s6sr27052f3yySiyXEF
sr2705.f3(() -> Swift.Int) -> ()
```

Reply:
Ah, it does make total sense that autoclosure is encoded in the mangled name, otherwise `func f(_: () -> Int)` and `func f(_: @autoclosure () -> Int)` would be mangled as the same which isn't right. 

> There is a swift-demangle utility shipped with toolchain/compiler

Awesome, thank you for the info =]